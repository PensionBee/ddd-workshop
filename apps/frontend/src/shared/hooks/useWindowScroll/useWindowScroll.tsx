import { useEffect, useState } from "react";

export const useWindowScroll = () => {
  const [scrollPosition, setScrollPosition] = useState(window.scrollY);
  useEffect(() => {
    const handleScroll = () => {
      setScrollPosition(window.scrollY);
    };
    window.addEventListener("scroll", handleScroll, { passive: true });
    return () => {
      window.removeEventListener("scroll", handleScroll);
    };
  }, []);
  return scrollPosition;
};
