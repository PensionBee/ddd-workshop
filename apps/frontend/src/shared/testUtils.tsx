import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
