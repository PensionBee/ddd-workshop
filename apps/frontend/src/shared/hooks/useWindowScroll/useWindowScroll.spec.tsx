import { act, render } from "@shared/testUtils";
import { vi } from "vitest";
import { useWindowScroll } from "./useWindowScroll";

const SCROLL_EVENT = new Event("scroll");

const JSDOM_DEFAULT_SCROLL_Y = "0";

const Component = () => {
  const scrollPosition = useWindowScroll();

  return <div data-testid="scroll-position" data-value={scrollPosition} />;
};

describe("useWindowScroll", () => {
  it("should set the window scroll on initial render", () => {
    const { asFragment, getByTestId } = render(<Component />);
    const scrollPosition = getByTestId("scroll-position");

    expect(scrollPosition).toHaveAttribute(
      "data-value",
      JSDOM_DEFAULT_SCROLL_Y
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should update the window scroll position on scroll", () => {
    const { asFragment, getByTestId } = render(<Component />);
    const scrollPosition = getByTestId("scroll-position");

    const newScrollPosition = 500;

    act(() => {
      global.scrollY = newScrollPosition;
      global.dispatchEvent(SCROLL_EVENT);
    });
    expect(scrollPosition).toHaveAttribute(
      "data-value",
      `${newScrollPosition}`
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
