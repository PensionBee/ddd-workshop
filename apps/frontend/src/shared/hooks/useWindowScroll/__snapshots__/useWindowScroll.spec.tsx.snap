// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useWindowScroll should set the window scroll on initial render 1`] = `
<DocumentFragment>
  <div
    data-testid="scroll-position"
    data-value="0"
  />
</DocumentFragment>
`;

exports[`useWindowScroll should update the window scroll position on scroll 1`] = `
<DocumentFragment>
  <div
    data-testid="scroll-position"
    data-value="500"
  />
</DocumentFragment>
`;
